homeassistant/components/esphome/entity.py

```diff
@@ -157,6 +161,9 @@ def __init__(
         assert entry_data.device_info is not None
         device_info = entry_data.device_info
         self._device_info = device_info
+        if object_id := entity_info.object_id:
+            # Use the object_id to suggest the entity_id
+            self.entity_id = f"{domain}.{device_info.name}_{object_id}"
```

Comment:
Need to check if object id can contain '-' and if we need to convert to _. Or it there are any other problem chars or if the entity code will do this for us.  

Can add some tests for this

Reply:
… must start with a letter and can end with numbers.

… must not have a space in the name.

… can not have special characters except the underscore (“_“).

… must not be a keyword.